Improve quick start docs (#172)
stm32-rs/stm32f1xx-hal: README.md

[![crates.io](https://img.shields.io/crates/v/stm32f1xx-hal.svg)](https://crates.io/crates/stm32f1xx-hal)
[![Released API docs](https://docs.rs/stm32f1xx-hal/badge.svg)](https://docs.rs/stm32f1xx-hal)

## Quick start guide

Embedded rust development requires a bit more setup than ordinary development.
For this guide, we'll assume you're using a stm32 blue pill board (shown
below), but if you have another f1 microcontroller, you should be able to adapt
it.

![blue pill pinout](BluePillPinout.jpg "opt title")

> You will also need an [stlink](https://www.adafruit.com/product/2548) for

Reviewer: This looks like a rebranded Chinesium ST-Link clone with pirated firmware (not sure how they can offer that on in the adafruit line). Can we recommend a genuine Product instead (like the ST-Link V3 Mini) which is even cheaper, though quite a bit trickier to obtain, insteads?

Reviewer: Alternatively we could recommend an open product or maybe hint to an official Nucleo board which happens to include an ST-Link so works out-of-the-box.

Author: I think we should use something compatible with blue pills since they seem pretty common in the rust community.

Recommending a better alternative to the weird stlink sounds like a good idea, got a link to one?

Reviewer: Any debug probe should be compatible with the blue pill. The question is only whether we want to stay with a native ST solution or recommend something else. The ST-Link v3 mini can be found here: https://www.st.com/en/development-tools/stlink-v3mini.html

### Installing software

To program your microcontroller, you need to install:
- [openocd](http://openocd.org/)
- `arm-none-eabi-gdb`

Finally, you need to install arm target support for the rust compiler. To do
so, run
```
rustup target install thumbv7m-none-eabi
```

### Setting up your project

Create a new rust project as you usually do with `cargo init`. The hello world
of embedded development is usually to blink an LED and code to do so is
available in [examples/blinky.rs](examples/blinky.rs). Copy that file to the
`main.rs` of your project.

You also need to add some dependencies to your `Cargo.toml`:

```toml
[dependencies]
embedded-hal = "0.2.3"
nb = "0.1.2"
cortex-m = "0.6.2"
cortex-m-rt = "0.6.11"
# Panic behaviour, see https://crates.io/keywords/panic-impl for alternatives
panic-halt = "0.2.0"

[dependencies.stm32f1xx-hal]
version = "0.5.2"
features = ["rt", "stm32f103"]
```

If you build your project now, you should get a single error: `error: language item required, but not found: eh_personality`
This unhelpful error message
can be fixed by compiling for the right target. To do so, run
```bash
cargo build --target thumbv7m-none-eabi
```

> 

Reviewer: Maybe also include a few words about putting that into a .cargo/config here? Can also do this as a follow up.

### Programming the microcontroller

If all went well, cargo should have successfully built your project. That means
it is time to actually run it on the hardware.  To do so plug your stlink into
the blue pill and start `openocd` using
```bash
openocd -f interface/stlink-v2.cfg -f target/stm32f1x.cfg
```
If you are not using an stlink V2, change the interface accordingly. 
For more information, see the [embeddonomicon].

If all went well, it should detect your microcontroller and say `Info :
stm32f1x.cpu: hardware has 6 breakpoints, 4 watchpoints`. Keep it running in
the background

We now need to tell `cargo` how to run the project and how to lay out the code
in memory. To do so, copy the following files from this repo to your project.

- [.gdbinit](.gdbinit)
- [.cargo/config](.cargo/config)

> - [memory.x](memory.x)

Reviewer: Uhm, that really should go in the section above as it is required (or at least useful in case of the .cargo/config) to build the software.

Author: Yea, that might be a good idea.

However, my point in ordering it like this is that the project successfully compiles after the cargo build line. If people get there and build fails, they have less things to check when debugging.

Reviewer: If you don't have a `memory.x` there's a good chance it will not compile or throw linker errors (depending on circumstances) so that should go before any attempts to compile it. `.cargo/config` is a wash but would simplify the change for error in the naked `cargo build` above.

[![crates.io](https://img.shields.io/crates/v/stm32f1xx-hal.svg)](https://crates.io/crates/stm32f1xx-hal)
[![Released API docs](https://docs.rs/stm32f1xx-hal/badge.svg)](https://docs.rs/stm32f1xx-hal)

## Quick start guide

Embedded rust development requires a bit more setup than ordinary development.
For this guide, we'll assume you're using a stm32 blue pill board (shown
below), but if you have another f1 microcontroller, you should be able to adapt
it.

![blue pill pinout](BluePillPinout.jpg "opt title")

You will also need a debug probe, for example an [stlink v3

> mini](https://www.adafruit.com/product/2548) for programming and debugging.

Reviewer: Wrong link

Author: Oops, fixed

> (There are many different stlinks out there, all of them _should_ work fine)

Reviewer: ```suggestion
(There are many different STLink probes out there, all of them _should_ work fine with this instructions given here, other JTAG or SWD debug probes will work as well but will need different software or configuration)
```

### Installing software

To program your microcontroller, you need to install:
- [openocd](http://openocd.org/)
- `arm-none-eabi-gdb`

Finally, you need to install arm target support for the rust compiler. To do
so, run
```
rustup target install thumbv7m-none-eabi
```

### Setting up your project

> Create a new rust project as you usually do with `cargo init`. The hello world

Reviewer: ```suggestion
Create a new Rust project as you usually do with `cargo init`. The hello world
```

of embedded development is usually to blink an LED and code to do so is
available in [examples/blinky.rs](examples/blinky.rs). Copy that file to the
`main.rs` of your project.

You also need to add some dependencies to your `Cargo.toml`:

```toml
[dependencies]
embedded-hal = "0.2.3"
nb = "0.1.2"
cortex-m = "0.6.2"
cortex-m-rt = "0.6.11"
# Panic behaviour, see https://crates.io/keywords/panic-impl for alternatives
panic-halt = "0.2.0"

[dependencies.stm32f1xx-hal]
version = "0.5.2"
features = ["rt", "stm32f103"]
```

If you build your project now, you should get a single error: `error: language

> item required, but not found: eh_personality` This unhelpful error message 

Reviewer: ```suggestion
item required, but not found: eh_personality`. This unhelpful error message 
```

> is fixed by compiling for the right target

Reviewer: ```suggestion
is fixed by compiling for the right target.
```

We also need to tell rust how to link our executable, and how to lay out the
result in memory. To accomplish this, copy [.cargo/config](.cargo/config) and
[memory.x](memory.x) from the stm32f1xx-hal repo to your project

```bash
cargo build
```

If everything went well, your project should have built without errors.

### Programming the microcontroller

It is now time to actually run the code on the hardware.  To do so plug your
stlink into the blue pill and start `openocd` using
```bash
openocd -f interface/stlink-v3.cfg -f target/stm32f1x.cfg
```
If you are not using an stlink V3, change the interface accordingly. 
For more information, see the [embeddonomicon].

If all went well, it should detect your microcontroller and say `Info :
stm32f1x.cpu: hardware has 6 breakpoints, 4 watchpoints`. Keep it running in

> the background

Reviewer: ```suggestion
the background.
```

mini](https://www.st.com/en/development-tools/stlink-v3mini.html) for programming and debugging.
(There are many different STLink probes out there, all of them _should_ work fine with this instructions given here, other JTAG or SWD debug probes will work as well but will need different software or configuration).

### Installing software

To program your microcontroller, you need to install:
- [openocd](http://openocd.org/)
- `arm-none-eabi-gdb`

Finally, you need to install arm target support for the Rust compiler. To do
so, run
```
rustup target install thumbv7m-none-eabi
```

### Setting up your project

Create a new Rust project as you usually do with `cargo init`. The hello world
of embedded development is usually to blink an LED and code to do so is
available in [examples/blinky.rs](examples/blinky.rs). Copy that file to the
`main.rs` of your project.

You also need to add some dependencies to your `Cargo.toml`:

```toml
[dependencies]
embedded-hal = "0.2.3"
nb = "0.1.2"
cortex-m = "0.6.2"
cortex-m-rt = "0.6.11"
# Panic behaviour, see https://crates.io/keywords/panic-impl for alternatives
panic-halt = "0.2.0"

[dependencies.stm32f1xx-hal]
version = "0.5.2"

> features = ["rt", "stm32f103"]

Reviewer 2: `features = ["rt", "stm32f103", "medium"]`
